File: 213_打家劫舍II/rob.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]

        dp1 = [0, nums[1]]
        for i in range(2, n):
            dp1.append(max(dp1[i-1], dp1[i-2] + nums[i]))

        dp2 = [nums[0], nums[0]]
        for i in range(2, n-1):
            dp2.append(max(dp2[i-1], dp2[i-2] + nums[i]))
        dp2.append(dp2[-1])

        return max(dp1[-1], dp2[-1])


# if __name__ == '__main__':
#     solu = Solution()
#     nums = [1,2,1,1]
#     print(solu.rob(nums))

File: 213_打家劫舍II/rob.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if(nums.length == 1) return nums[0]
    function robrange(nums, start, end){
        let cur = 0; let pos1 = 0; let pos2 = 0
        for(let i = end-1; i >= start; i--){
            cur = Math.max(pos1, pos2+nums[i])
            pos2 = pos1
            pos1 = cur
        }
        return cur
    }
    return Math.max(robrange(nums, 0, nums.length-1), robrange(nums, 1, nums.length))
};